src/nvim/screen.c

```diff
@@ -3949,40 +3971,84 @@ win_line (
             && (int)wp->w_virtcol <
             wp->w_width * (row - startrow + 1) + v
             && lnum != wp->w_cursor.lnum)
-           || draw_color_col)
-          && !wp->w_p_rl
-          ) {
+           || draw_color_col || do_eoltext)
+          && !wp->w_p_rl) {
         int rightmost_vcol = 0;
         int i;
+        size_t eol_pos = 0;
+        LineState s = LINE_STATE((char_u *)"");
+        int eol_attr = 0;
+
+        VirtText virt_text = do_eoltext ? bufhl_info.line->virt_text
+                                        : (VirtText)KV_INITIAL_VALUE;
+        bool delay_text = lcs_eol <= 0;
 
-        if (wp->w_p_cuc)
+        if (wp->w_p_cuc) {
           rightmost_vcol = wp->w_virtcol;
-        if (draw_color_col)
-          /* determine rightmost colorcolumn to possibly draw */
-          for (i = 0; color_cols[i] >= 0; ++i)
-            if (rightmost_vcol < color_cols[i])
+        }
+
+        if (draw_color_col) {
+          // determine rightmost colorcolumn to possibly draw
+          for (i = 0; color_cols[i] >= 0; i++) {
+            if (rightmost_vcol < color_cols[i]) {
               rightmost_vcol = color_cols[i];
+            }
+          }
+        }
 
         int cuc_attr = win_hl_attr(wp, HLF_CUC);
         int mc_attr = win_hl_attr(wp, HLF_MC);
 
         while (col < wp->w_width) {
-          schar_from_ascii(ScreenLines[off], ' ');
-          col++;
+          int cells = -1;
+          if (do_eoltext && !delay_text) {
+            if (*s.p == NUL) {
+              if (eol_pos < virt_text.size) {
+                s.p = (char_u *)kv_A(virt_text, eol_pos).text;
+                int hl_id = kv_A(virt_text, eol_pos).hl_id;
+                eol_attr = hl_id > 0 ? syn_id2attr(hl_id) : 0;
+                eol_pos++;
+              } else {
+               do_eoltext = false;
+              }
+            }
+            if (*s.p != NUL) {
+              cells = line_putchar(&s, &ScreenLines[off], wp->w_width - col,
+                                   false);
+            }
+          }
+          delay_text = false;
```

Comment:
If `do_eoltext` were a TriState then this `if ...` block could be:

```c
if (do_eoltext == kTrue) {
  ...
    do_eoltext = kFalse;
} else if (do_eoltext == kNone) {
  do_eoltext = kTrue;
}
```

This doesn't help much ... but there are a lot of new flags being introduced, and  it's one less symbol to keep track of mentally ...

Reply:
TriState doesn't represent the state well,  `delay_virttext` sounds simple enough to me, but otherwise I would rather create a new specific enum (to not just replace "boolean blindness" with "trinary blindness")